### ISSUE TITLE
'%s' % Unknown should be str

### ISSUE THREAD

**Issue by eagleoflqj** [microsoft/pyright#3462]
**Describe the bug**
`'%s' % a` is deduced `Unknown` when `a` is not type-hinted.

**To Reproduce**
```py
def f(a):
    reveal_type('{}'.format(a))
    reveal_type('%s' % a)
```
**Expected behavior**
Both are str.

**VS Code extension or command-line**
Command line 1.1.247.

**Comment by erictraut** [comment]
The parameter `a` has no type annotation, so its type is inferred to be `Unknown`. The expression `'%s' % a` is therefore also `Unknown`.

**Comment by eagleoflqj** [comment]
I make a class A that overrides __rmod__ to always return 0 regardless what the parameter is, and call '%s'%A() and I still get a str. If the result can only be str, why not make it str?
Add type hints to parameter is not always feasible as they are hard to be annotated accurately sometimes.

**Comment by erictraut** [comment]
If you are assuming that `a` is an instance of class `A`, you should annotate the parameter accordingly.

Binary operations in Python are complex operations that involve the evaluation of multiple magic methods. For example, the `%` operator maps to `__mod__` and `__rmod__`. A type checker needs to understand the types of both the left-hand and right-hand operands to properly evaluate the type of the expression. It needs to handle additional complexities like union types, generic classes and methods, and overloaded method signatures. If the type of the LHS or RHS is unknown, it cannot properly apply the appropriate type rules.
